ios/LayoutReanimation/REASharedTransitionManager.m

```diff
@@ -0,0 +1,373 @@
+#import <objc/runtime.h>
+
+@implementation REASharedTransitionManager {
+  NSMutableDictionary<NSNumber *, UIView *> *_sharedTransitionParent;
+  NSMutableDictionary<NSNumber *, NSNumber *> *_sharedTransitionInParentIndex;
+  NSMutableDictionary<NSNumber *, REASnapshot *> *_snapshotRegistry;
+  NSMutableArray<UIView *> *_currentSharedTransitionViews;
+  REAFindTheOtherForSharedTransitionBlock _findTheOtherForSharedTransition;
+  UIView *_transitionContainer;
+  NSMutableArray<UIView *> *_addedSharedViews;
+  BOOL _isSharedTransitionActive;
+  NSMutableArray<REASharedElement *> *_sharedElements;
+  REAAnimationsManager *_animationManager;
+}
+
+- (instancetype)init:(REAAnimationsManager *)animationManager
+{
+  if (self = [super init]) {
+    _snapshotRegistry = [NSMutableDictionary new];
+    _currentSharedTransitionViews = [NSMutableArray new];
+    _addedSharedViews = [NSMutableArray new];
+    _sharedTransitionParent = [NSMutableDictionary new];
+    _sharedTransitionInParentIndex = [NSMutableDictionary new];
+    _isSharedTransitionActive = NO;
+    _sharedElements = [NSMutableArray new];
+    _animationManager = animationManager;
+  }
+  return self;
+}
+
+- (void)invalidate
+{
+  _snapshotRegistry = nil;
+  _currentSharedTransitionViews = nil;
+  _addedSharedViews = nil;
+  _sharedTransitionParent = nil;
+  _sharedTransitionInParentIndex = nil;
+  _sharedElements = nil;
+  _animationManager = nil;
+}
+
+- (NSArray<UIView *> *)getCurrentSharedTransitionViews
+{
+  return _currentSharedTransitionViews;
+}
+
+- (void)notifyAboutNewView:(UIView *)view
+{
+  [_addedSharedViews addObject:view];
+}
+
+- (void)viewsDidLayout
+{
+  [self setupAsyncSharedTransitionForViews:_addedSharedViews];
+  [_addedSharedViews removeAllObjects];
+}
+
+- (void)setupAsyncSharedTransitionForViews:(NSArray<UIView *> *)views
+{
+  NSArray *sharedViews = [self sortViewsByTags:views];
+  _sharedElements = [self getSharedElementForCurrentTransition:sharedViews withNewElements:YES];
+}
+
+- (void)setupSyncSharedTransitionForViews:(NSArray<UIView *> *)views
+{
+  NSArray *sharedViews = [self sortViewsByTags:views];
+  NSArray<REASharedElement *> *sharedElements = [self getSharedElementForCurrentTransition:sharedViews
+                                                                           withNewElements:NO];
+  if ([sharedElements count] == 0) {
+    return;
+  }
+  [self setupTransitionContainer];
+  [self reparentSharedViewsForCurrentTransition:sharedElements];
+  [self startSharedTransition:sharedElements];
+}
+
+- (NSArray *)sortViewsByTags:(NSArray *)views
+{
+  return [views sortedArrayUsingComparator:^NSComparisonResult(UIView *view1, UIView *view2) {
+    if ([view1.reactTag intValue] > [view2.reactTag intValue]) {
+      return (NSComparisonResult)NSOrderedAscending;
+    }
+    if ([view1.reactTag intValue] < [view2.reactTag intValue]) {
+      return (NSComparisonResult)NSOrderedDescending;
+    }
+    return (NSComparisonResult)NSOrderedSame;
```

Comment:
Why not use NSNumber's `compare:`? https://developer.apple.com/documentation/foundation/nsnumber/1413562-compare?language=objc ?